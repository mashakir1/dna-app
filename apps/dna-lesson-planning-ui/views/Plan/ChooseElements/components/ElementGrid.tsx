// Components
import { ElementCard } from "components/Cards/ElementCard";

// Types
import { Element } from "types";

interface ElementGridProps {
  elements: Array<Element>;
  selectedIds: Array<string>;
  addSelectedId: (id: string) => void;
  removeSelectedId: (id: string) => void;
}

const ElementGrid: (props: ElementGridProps) => JSX.Element = (
  { elements, selectedIds, addSelectedId, removeSelectedId }
) => {
  const elementOnClick = (selectedElementId: string) => {
    if (selectedIds.includes(selectedElementId)) {
      removeSelectedId(selectedElementId);
    } else {
      addSelectedId(selectedElementId);
    }
  };

  return <div className={"flex gap-m wrap"}>
    {elements?.map((element: Element) => (
      <ElementCard
        key={element.id}
        id={element.id}
        title={element.subject}
        text={element.title}
        subjectColour={element.subjectColour}
        duration={element.duration}
        onClick={elementOnClick}
        selected={selectedIds.includes(element.id)}
      />
    ))}
  </div>;
};

export default ElementGrid;
