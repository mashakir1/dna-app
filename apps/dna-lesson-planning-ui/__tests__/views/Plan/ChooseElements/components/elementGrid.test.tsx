import { fireEvent, render, screen } from "@testing-library/react";
import ElementGrid from "views/Plan/ChooseElements/components/ElementGrid";
import { testElements } from "../mockData";

jest.mock("components/Cards/ElementCard", () => ({
  ElementCard({ id, onClick }) {
    return <div data-testid={id} onClick={() => onClick(id)} />;
  }
}));

describe("ElementGrid", () => {
  it("empty grid matches snapshot", () => {
    const { container } = render(<ElementGrid
      elements={[]} selectedIds={[]}
      addSelectedId={() => undefined} removeSelectedId={() => undefined}
    />);
    expect(container).toMatchSnapshot();
  });

  it("populated grid matches snapshot", () => {
    const { container } = render(<ElementGrid
      elements={testElements} selectedIds={[]}
      addSelectedId={() => undefined} removeSelectedId={() => undefined}
    />);
    expect(container).toMatchSnapshot();
  });

  it("adds elements to selection correctly using callback", () => {
    const testAddFn = jest.fn();
    const testRemoveFn = jest.fn();

    render(<ElementGrid
      elements={testElements} selectedIds={[]}
      addSelectedId={testAddFn} removeSelectedId={testRemoveFn}
    />);

    const testElementCard = screen.queryByTestId("testId1");
    expect(testElementCard).toBeTruthy();
    fireEvent.click(testElementCard);
    expect(testAddFn).toHaveBeenCalled();
    expect(testRemoveFn).not.toHaveBeenCalled();
  });

  it("removes elements to selection correctly using callback", () => {
    const testAddFn = jest.fn();
    const testRemoveFn = jest.fn();

    render(<ElementGrid
      elements={testElements} selectedIds={["testId1"]}
      addSelectedId={testAddFn} removeSelectedId={testRemoveFn}
    />);

    const testElementCard = screen.queryByTestId("testId1");
    expect(testElementCard).toBeTruthy();
    fireEvent.click(testElementCard);
    expect(testAddFn).not.toHaveBeenCalled();
    expect(testRemoveFn).toHaveBeenCalled();
  });
});
