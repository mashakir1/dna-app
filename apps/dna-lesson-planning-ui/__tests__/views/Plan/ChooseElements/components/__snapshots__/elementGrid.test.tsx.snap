// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElementGrid empty grid matches snapshot 1`] = `
<div>
  <div
    class="flex gap-m wrap"
  />
</div>
`;

exports[`ElementGrid populated grid matches snapshot 1`] = `
<div>
  <div
    class="flex gap-m wrap"
  >
    <div
      data-testid="testId1"
    />
    <div
      data-testid="testId2"
    />
    <div
      data-testid="testId3"
    />
  </div>
</div>
`;
